File: detect cycle in undirected graph/has_cycle.go
package main

import "fmt"

func hasCycle(numNodes int, edges [][]int) bool {
	graph := buildGraph(numNodes, edges)
	fmt.Printf("adjacency list: %+v\n\n", graph)

	visited := map[int]bool{}

	for node := 0; node < numNodes; node += 1 {
		fmt.Println("------------------------------------------")
		fmt.Println("dfs node ", node)
		if detectCycle(graph, node, visited, -1) {
			return false
		}
	}

	return true
}

func detectCycle(
	graph map[int][]int,
	node int,
	visited map[int]bool,
	parent int) bool {
	if _, found := visited[node]; found {
		fmt.Printf("node %d has been visited\n\n", node)
		return true
	}

	visited[node] = true
	for _, descendant := range graph[node] {
		fmt.Printf("current node: %d\n", node)
		fmt.Printf("visited: %+v\n", visited)
		fmt.Printf("parent: %d\n", parent)
		if descendant != parent {
			fmt.Printf("visited descendant: node %d\n\n", descendant)
		} else {
			fmt.Printf("not visited descendant, node %d. It is parent.\n\n", descendant)
		}
		if descendant != parent && detectCycle(graph, descendant, visited, node) {
			return true
		}
	}
	return false
}

func buildGraph(numNodes int, prerequisites [][]int) map[int][]int {
	graph := map[int][]int{}
	for i := 0; i < numNodes; i++ {
		graph[i] = []int{}
	}

	for _, prerequisite := range prerequisites {
		a, b := prerequisite[0], prerequisite[1]
		graph[b] = append(graph[b], a)
		graph[a] = append(graph[a], b)
	}

	return graph
}

func main() {
	numCourses := 6
	prerequisites := [][]int{
		{1, 0},
		{3, 0},
		{5, 0},
		{2, 1},
		{3, 1},
		{4, 3},
		{5, 4},
		{3, 5},
	}
	fmt.Println("has cycle?", hasCycle(numCourses, prerequisites))
}

/*
output:
adjacency list: map[0:[1 3 5] 1:[0 2 3] 2:[1] 3:[0 1 4 5] 4:[3 5] 5:[0 4 3]]

------------------------------------------
dfs node  0
current node: 0
visited: map[0:true]
parent: -1
visited descendant: node 1

current node: 1
visited: map[0:true 1:true]
parent: 0
not visited descendant, node 0. It is parent.

current node: 1
visited: map[0:true 1:true]
parent: 0
visited descendant: node 2

current node: 2
visited: map[0:true 1:true 2:true]
parent: 1
not visited descendant, node 1. It is parent.

current node: 1
visited: map[0:true 1:true 2:true]
parent: 0
visited descendant: node 3

current node: 3
visited: map[0:true 1:true 2:true 3:true]
parent: 1
visited descendant: node 0

node 0 has been visited

has cycle? false
*/

File: detect cycle in undirected graph/has_cycle.py
def hasCycle(numNodes, edges):
    graph = buildGraph(numNodes, edges)
    print("adjacency list", graph)
    print("")
    visited = set()
    for node in range(numNodes):
        print("----------------------------------------------")
        print("dfs node", node)
        if node not in visited and detectCycle(graph, node, visited, None):
            return True
    
    return False
        
def detectCycle(graph, node, visited, parent):
    if node in visited:
        print("node ", node, "has been visited")
        return True
    
    visited.add(node)
    for descendant in graph[node]:
        print("current node: node", node)
        print("visited", visited)
        print("parent: node", parent)
        print("visit descendant: node", descendant) if descendant != parent else "not visit descendant node because it is parent"
        print("")
        if descendant != parent and detectCycle(graph, descendant, visited, node):
            return True

    return False
    
    
def buildGraph(numNodes, edges):
    graph = {x: [] for x in range(numNodes)}
    for edge in edges:
        a, b = edge
        # bidirection
        graph[b].append(a)
        graph[a].append(b)
    return graph
        

if __name__ == "__main__":
    numNodes = 6
    edges = [[1, 0], [3, 0], [5, 0], [2, 1], [3, 1], [4, 3], [5, 4], [3, 5]]
    print("has cycle?", hasCycle(numNodes, edges))

"""
output:
('adjacency list', {0: [1, 3, 5], 1: [0, 2, 3], 2: [1], 3: [0, 1, 4, 5], 4: [3, 5], 5: [0, 4, 3]})

----------------------------------------------
('dfs node', 0)
('current node: node', 0)
('visited', set([0]))
('parent: node', None)
('visit descendant: node', 1)

('current node: node', 1)
('visited', set([0, 1]))
('parent: node', 0)
not visit descendant node because it is parent

('current node: node', 1)
('visited', set([0, 1]))
('parent: node', 0)
('visit descendant: node', 2)

('current node: node', 2)
('visited', set([0, 1, 2]))
('parent: node', 1)
not visit descendant node because it is parent

('current node: node', 1)
('visited', set([0, 1, 2]))
('parent: node', 0)
('visit descendant: node', 3)

('current node: node', 3)
('visited', set([0, 1, 2, 3]))
('parent: node', 1)
('visit descendant: node', 0)

('node ', 0, 'has been visited')
('has cycle?', True)
"""